### PR TITLE
Updates to documentation on contributing to gist. Fixes #382.

### LINKED ISSUE #382: Updates to Contributing.md
Add to Contributing.md:

- Notes on assigning a PR to a release
- Clarify the case of multiple developers working off one main development branch
- Miscellaneous wordsmithing

### PR DESCRIPTION
**Pull request by rjyounes**

### CODE REVIEW

File: docs/Contributing.md [hunk 1]
```diff
@@ -56,50 +56,48 @@ Submitting an Issue
 Implementation
 -----
 
-### Development Branch
+Note: The following instructions reference two git branches: the working branch, which is the one you create for your development work, and the main development branch, which is the branch that your working branch is cut from and will be merged back into. In the majority of cases the latter is the default "develop" branch, but there are instances, described below, where it is not, so the term "main development branch" is used instead.
```
Comment (uscholdm) [inline docs/Contributing.md:59]: If there are two foobar things one is a plain foobar and the other is the main foobar, then noone will think that the main foobar is subordinate to the plain foobar.  

Thus, it's intuitively jarring that the 'main development branch' would be subordinate to a plain 'develop' branch.  Here is an alternative: 

However many levels deep into branching and sub-branching: 
1. Let's call any branch on which the main work is meant to take place a 'working branch'. 
2. Let's call any branch that a working branch branches off of and merges back into a 'develop branch'.

The develop branch that branches off the master can be called the 'main develop branch'. 
A develop branch that branches off of the main develop branch can be called a 'local develop branch'.

This is a perspective thing, not an absolute thing.  There is nothing to stop one from doing work on what someone else is treating as a local develop branch.  One man's develop branch can be another's working branch.

File: docs/Contributing.md [hunk 2]
```diff
@@ -56,50 +56,48 @@ Submitting an Issue
...
 - However, it is possible to go overboard and commit every little change independently. This creates clutter in the repository history.
-- This is especially important if working on a project with others, to reduce the likelihood of conflicts.
-- However, each commit should be atomic for ease of rollback or reversion. Finish working on one sub-task and commit it before taking up another.
-- As you work, it is _essential_ that you merge regularly from develop back into your branch. This ensures that, when it comes time to merge your work into develop, you will have resolved any merge conflicts with a minimum of difficulty.
+- As you work, it is _essential_ that you merge regularly from the main development branch back into your working branch. This ensures that, when it comes time to merge your work into the main branch, you will have resolved any merge conflicts with a minimum of difficulty. Note that your PR cannot be merged to the main branch until all merge conflicts are resolved.
```
Comment (uscholdm) [inline docs/Contributing.md:76]: > PR cannot be merged to the main branch

Do you mean 'master' branch here rather than main branch?

File: docs/Contributing.md [hunk 3]
```diff
@@ -56,50 +56,48 @@ Submitting an Issue
...
 Pull Requests
 -----
 
 ### Creating the Pull Request (PR)
 
-- Once your work is ready to be merged into the develop branch, you will create a pull request (PR).
-- Before submitting the PR, you should ensure that you have run the serializer and pulled from develop into your working branch, as above for implementation.
-- Submit the PR to develop (the default branch) and assign it to the release project.
-- Assign one or more reviewers, as specified below.
+- Once your work is ready to be merged into the main development branch, you will create a pull request (PR).
```
Comment (uscholdm) [inline docs/Contributing.md:88]: > Once your work is ready to be merged into the main development branch

I guess this means what I would call a local develop branch for the project, or 'project develop branch'.

File: docs/Contributing.md [hunk 4]
```diff
@@ -129,25 +127,26 @@ Release notes gist 9.4.0
...
+- As well as designating reviewers, you should assign the PR &mdash; but not the issue &mdash; to the reviewers. This makes the assignee visible on the project board so it is clear who is currently responsible for moving the PR forward.
 
 ### Review Process
 
 - Reviewers are charged with rejecting the PR or requesting changes, with appropriate comments, if:
   - The criteria of atomicity and completeness are not met.
   - The implementation deviates from that agreed upon during the internal review.
-- If changes are requested, reassign the issue to the original implementer so it is clear who is responsible for the next action.
+  - The release notes have been updated.
+- A reviewer should assign the PR to the same release project as the issue it addresses if that has not been done when the PR was submitted.
```
Comment (uscholdm) [inline docs/Contributing.md:141]: Is there an assumption that on PR addresses exactly one issue?

**Review by uscholdm — CHANGES_REQUESTED** [review]
I found the terminology a bit confusing and after a bit of head-scratching came up with an alternative.

File: docs/Contributing.md (hunk 1 above)
Comment (rjyounes) [reply docs/Contributing.md:59]: I agree that the terminology I've used is confusing, though I don't think you've quite hit the mark in your proposal. For one thing, "develop branch" is bad grammar: don't form compounds with verb + noun, only adjective or noun + noun; e.g., we don't way "organize committee" but rather "organizing committe." "Develop" has become a conventional name for a particular branch in the repository that plays a particular function in the overall scheme. I'll reword this and run it by you again.

Actually, there may be branch rules in place that prevent others from developing on (committing to) a particular branch. Currently we have such rules in place for master and develop, but one could just as well create rules on a branch off of develop.

File: docs/Contributing.md (hunk 2 above)
Comment (rjyounes) [reply docs/Contributing.md:76]: No. Git will not allow you to merge one branch into another until merge conflicts have been resolved, unless the branch allows force commits, which is a _very_ bad idea.

File: docs/Contributing.md (hunk 4 above)
Comment (rjyounes) [reply docs/Contributing.md:141]: No. Will fix.

File: docs/Contributing.md (hunk 1 above)
Comment (uscholdm) [reply docs/Contributing.md:59]: > For one thing, "develop branch" is bad grammar

If the branch called 'develop' was called 'developing' I would have used that.  A matter of taste, but to me, making a direct connection to what is super common thus adding clarity, trumps a grammatical nit that few would even notice (present company excepted :-)  I await a new proposal.

Comment (uscholdm) [reply docs/Contributing.md:59]: > "Local" implies that the branch is not pushed to the remote

Good point, that won't work.

File: docs/Contributing.md (hunk 2 above)
Comment (uscholdm) [reply docs/Contributing.md:76]: OK so clarity is needed on the meaning of "main branch".  Is it shorthand for "man development branch"?

**Comment by rjyounes** [comment]
@uscholdm I decided it was too confusing to describe the two cases at once (it seems my developer impulses were at work, assigning variables and using a recursive function), and at any rate the multi-developer scenario is something of an edge case, or at least not common in our practice. So I cordoned off that discussion to its own small section. See what you think.

File: docs/Contributing.md [hunk 5]
```diff
@@ -56,50 +56,43 @@ Submitting an Issue
...
 - However, it is possible to go overboard and commit every little change independently. This creates clutter in the repository history.
-- This is especially important if working on a project with others, to reduce the likelihood of conflicts.
-- However, each commit should be atomic for ease of rollback or reversion. Finish working on one sub-task and commit it before taking up another.
-- As you work, it is _essential_ that you merge regularly from develop back into your branch. This ensures that, when it comes time to merge your work into develop, you will have resolved any merge conflicts with a minimum of difficulty.
+- As you work, it is _essential_ that you merge regularly from develop back into your working branch. This ensures that, when it comes time to merge your work into develop, you will have resolved any merge conflicts with a minimum of difficulty. Note that your PR cannot be merged to develop until all merge conflicts are resolved.
```
Comment (uscholdm) [inline docs/Contributing.md:72]: I rarely do that because most of the changes I do are small and done over the course of a few hours. This is only important if plenty of other work is going on at the same time which is more likely if the work is taking place over a longerish period of time (such as now with skos-annotations).

Also, does it make it harder to see what the changes were for a particular branch?  If you do a big task over 2 weeks and merge develop into your branch 3 times for 3 other pieces of work, is there still an easy way for me to look at the work that was ONLY done on your branch?  Probably there is, just wondering.

File: docs/Contributing.md [hunk 6]
```diff
@@ -56,50 +56,43 @@ Submitting an Issue
...
 -----
 
 ### Creating the Pull Request (PR)
 
-- Once your work is ready to be merged into the develop branch, you will create a pull request (PR).
-- Before submitting the PR, you should ensure that you have run the serializer and pulled from develop into your working branch, as above for implementation.
-- Submit the PR to develop (the default branch) and assign it to the release project.
-- Assign one or more reviewers, as specified below.
+- Once your work is ready to be merged into develop, you will create a pull request (PR).
+- Before submitting the PR, you should ensure that you have (a) run the serializer and (b) merged develop into your working branch, as above [during implementation](#commits-pushes-and-merges).
```
Comment (uscholdm) [inline docs/Contributing.md:84]: If following instructions, it should not be necessary to run serialzer AGAIN. So maybe reword this to emphasize that? Small point, feel free to ignore.

> Before submitting the PR, you should ensure that you have merged develop into your working branch, as above. Every commit should have the serializer fun, but just in case, makes double sure.

File: docs/Contributing.md [hunk 7]
```diff
@@ -56,50 +56,43 @@ Submitting an Issue
...
-- For a large project implemented by more than one person, all developers would work on this same branch. This applies whether the contributions of different developers are concurrent or sequential.
+- The first step is to create a working branch from develop for your work: e.g., `feature/document_submission_guidelines` or `bugfix/fix_typo_in_Address_definition`. This branch will not be preserved after the work is merged into develop, so its name does not have to be particularly specific, accurate, or detailed. Some developers like to add the issue number to the new branch.
 
 ### Style Guide
 
 - Ontologists should consult the emerging [_gist Style Guide_](gistStyleGuide.md) during implementation.
-- The serializer tool must be run before each commit in order to standardize formatting and eliminate noise in git diffs. See [_gist Style Guide_](gistStyleGuide.md) for details.
+- The serializer tool should be run before each commit in order to standardize formatting and eliminate noise in git diffs. See [_gist Style Guide_](gistStyleGuide.md) for details.
```
Comment (uscholdm) [inline docs/Contributing.md:66]: Not 'should', when is optional. The ontologist may choose to have this happen automatically as part of a commit if they done do it before the commit. That should be encourage, even enforced to the extent that we can. People (including me) often don't get around to this stuff.

Come to think of it, there is a good reason to a commit WITHOUT running serializer, and then follow with a commit immediately after running serializer. I often do this. Why?  Its far easier to look at file differences w/o the alpha sorting making a mess. This is a argument for not having the serializer auto-run (hook) or to have it be optionally turned off. Maybe too fiddly.

File: docs/Contributing.md [hunk 8]
```diff
@@ -56,50 +56,43 @@ Submitting an Issue
...
 
 ### Creating the Pull Request (PR)
 
-- Once your work is ready to be merged into the develop branch, you will create a pull request (PR).
-- Before submitting the PR, you should ensure that you have run the serializer and pulled from develop into your working branch, as above for implementation.
-- Submit the PR to develop (the default branch) and assign it to the release project.
-- Assign one or more reviewers, as specified below.
+- Once your work is ready to be merged into develop, you will create a pull request (PR).
+- Before submitting the PR, you should ensure that you have (a) run the serializer and (b) merged develop into your working branch, as above [during implementation](#commits-pushes-and-merges).
+- Submit the PR to develop.
```
Comment (uscholdm) [inline docs/Contributing.md:85]: > Submit the PR to develop

Reword to:
"Submit the PR to merge your working branch into develop"

**Review by uscholdm — CHANGES_REQUESTED** [review]
The language on the complicated project work is very clear now - better than either of our prior attempts.

A few suggested changes that you are free to disagree with...

File: docs/Contributing.md (hunk 8 above)
Comment (rjyounes) [reply docs/Contributing.md:85]: A PR is literally a "pull request" to pull the changes on one branch into another. So "pull request to merge" is redundant.

File: docs/Contributing.md (hunk 6 above)
Comment (rjyounes) [reply docs/Contributing.md:84]: Added something similar.

File: docs/Contributing.md (hunk 7 above)
Comment (rjyounes) [reply docs/Contributing.md:66]: We talked about this in the review and I respectfully disagree. You may preview your changes without the serializer, but I still think the serializer should be run before each commit. If you were working on a serialized version, there should not be many alpha sort issues. The one you showed us was an edge case, IMO.

File: docs/Contributing.md (hunk 5 above)
Comment (rjyounes) [reply docs/Contributing.md:72]: By "regularly" I didn't mean every 5 minutes :)  I've amended the instruction to suggest daily merges.

I guess to see only the work done on that branch you'd have to look at the commits before each merge. Admittedly somewhat inconvenient, but less so that the alternative. The alternative is to use rebase instead of merge, which keeps the history clean, but does so by rewriting the history. Rebase can be complicated and we (Boris, MarkW, and I) thought that it shouldn't be recommended practice.
